certification/internal/k8s/deployable_by_olm.go

```diff
@@ -0,0 +1,216 @@
+
+type DeployableByOlmCheck struct{}
+
+func (p *DeployableByOlmCheck) Validate(image string) (bool, error) {
+
+	// create k8s custom resources for the operator deployment
+	err = p.setUp()
+	defer p.cleanUp()
+	if err != nil {
+		return false, err
+	}
+	installedCSV, err := p.installedCSV()
+	if err != nil {
+		return false, err
+	}
+	return p.isCSVReady(installedCSV)
+}
+
+func (p *DeployableByOlmCheck) setUp() error {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	if len(kubeconfig) > 0 {
+		k8sconfig, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	catalogImage = viper.GetString("catalogImage")
+	channel = viper.GetString("channel")
+	packageName = viper.GetString("package")
+	app = viper.GetString("appName")
+	ooNamespace = viper.GetString("installNamespace")
+
+	_, err = openshiftEngine.CreateNamespace(ooNamespace, cli.OpenshiftOptions{}, k8sconfig)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	_, err = openshiftEngine.CreateCatalogSource(cli.CatalogSourceData{Name: app, Image: catalogImage}, cli.OpenshiftOptions{Namespace: catalogSourceNS}, k8sconfig)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	_, err = openshiftEngine.CreateOperatorGroup(cli.OperatorGroupData{Name: app, TargetNamespaces: targetNamespaces}, cli.OpenshiftOptions{Namespace: ooNamespace}, k8sconfig)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	subscriptionData := cli.SubscriptionData{
+		Name:                   app,
+		Channel:                channel,
+		CatalogSource:          app,
+		CatalogSourceNamespace: catalogSourceNS,
+		Package:                packageName,
+	}
+	_, err = openshiftEngine.CreateSubscription(subscriptionData, cli.OpenshiftOptions{Namespace: ooNamespace}, k8sconfig)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
+}
+
+func (p *DeployableByOlmCheck) isCSVReady(installedCSV string) (bool, error) {
+
+	var csv *operatorv1alpha1.ClusterServiceVersion
+
+	log.Trace(fmt.Sprintf("Looking for csv %s in namespace %s", installedCSV, ooNamespace))
+
+	// query API server for CSV by name
+	ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+	defer cancel()
+
+	for {
+		log.Debug("Waiting for ClusterServiceVersion to become ready...")
+		csv, _ = openshiftEngine.GetCSV(installedCSV, cli.OpenshiftOptions{Namespace: ooNamespace}, k8sconfig)
+		// if the CSV phase is succeeded or the context deadline exceeds, stop the querying
+		if csv.Status.Phase == operatorv1alpha1.CSVPhaseSucceeded {
+			log.Debug("CSV is created successfully: ", installedCSV)
+			break
+		}
+		// if the context deadline exceeds, fail the check
+		if ctx.Err() != nil {
+			log.Error(fmt.Sprintf("failed to fetch the csv: %s ", installedCSV), ctx.Err())
+			return false, errors.ErrK8sAPICallFailed
+		}
+		log.Debug("CSV is not ready yet, retrying...")
+		time.Sleep(2 * time.Second)
+	}
+	return true, nil
```

Comment:
Can we restructure this logic such that we explicitly find cases that return `true`, and otherwise return false?

We want to continue to explicitly find success cases, and otherwise return a failure to reduce false positives.

Reply:
Fixed...